File: PlusEduc-UI/src/services/studentPortal.ts
import { apiClient } from './api';
import type {
  Classmate,
  GradeRecord,
  QuestionAnswer,
  StudentActivity,
  StudentActivityDetail,
  StudentClassroom,
  StudentPortalPerformance,
  StudentPortalProfile,
  StudentSubmissionResult,
  TeacherSummary,
} from '@/types';

class StudentPortalService {
  async getProfile(): Promise<StudentPortalProfile> {
    return apiClient.get<StudentPortalProfile>('/student-portal/me');
  }

  async getClassroom(): Promise<StudentClassroom> {
    return apiClient.get<StudentClassroom>('/student-portal/classroom');
  }

  async getClassmates(): Promise<Classmate[]> {
    return apiClient.get<Classmate[]>('/student-portal/classmates');
  }

  async getTeachers(): Promise<TeacherSummary[]> {
    return apiClient.get<TeacherSummary[]>('/student-portal/teachers');
  }

  async getActivities(): Promise<StudentActivity[]> {
    return apiClient.get<StudentActivity[]>('/student-portal/activities');
  }

  async getActivity(id: string): Promise<StudentActivityDetail> {
    return apiClient.get<StudentActivityDetail>(`/student-portal/activities/${id}`);
  }

  async submitActivity(id: string, answers: QuestionAnswer[]): Promise<StudentSubmissionResult> {
    return apiClient.post<StudentSubmissionResult>(`/student-portal/activities/${id}/submissions`, { answers });
  }

  async getGrades(): Promise<GradeRecord[]> {
    return apiClient.get<GradeRecord[]>('/student-portal/grades');
  }

  async getPerformance(): Promise<StudentPortalPerformance> {
    return apiClient.get<StudentPortalPerformance>('/student-portal/performance');
  }

  async downloadActivityPdf(id: string, title: string): Promise<void> {
    const response = await fetch(`/api/student-portal/activities/${id}/export-pdf`, {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${localStorage.getItem('accessToken')}`,
      },
    });

    if (!response.ok) {
      throw new Error('Erro ao baixar PDF da atividade');
    }

    const blob = await response.blob();
    const url = window.URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = url;
    link.download = `${title.replace(/[^a-z0-9]/gi, '_').toLowerCase()}.pdf`;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
  }
}

export const studentPortalService = new StudentPortalService();

File: PlusEduc-BE-Python/app/api/student_portal.py
from fastapi import APIRouter, Depends, Request, Response

from app.core.auth import UserPrincipal
from app.dependencies import require_roles
from app.schemas.student_portal_schema import (
    Classmate,
    StudentActivity,
    StudentActivityDetail,
    StudentClassroom,
    StudentPortalGrade,
    StudentPortalPerformance,
    StudentPortalProfile,
    TeacherSummary,
)
from app.services.student_portal_service import StudentPortalService

router = APIRouter(prefix="/api/student-portal", tags=["Student Portal"])


def service(request: Request) -> StudentPortalService:
    return StudentPortalService(
        request.app.state.student_repository,
        request.app.state.classroom_repository,
        request.app.state.teacher_repository,
        request.app.state.activity_repository,
        request.app.state.activity_submission_repository,
        request.app.state.grade_repository,
        request.app.state.mongo,
    )


def student_user(current_user: UserPrincipal = Depends(require_roles("STUDENT"))) -> UserPrincipal:
    return current_user


@router.get("/me", response_model=StudentPortalProfile)
def get_profile(request: Request, current_user: UserPrincipal = Depends(student_user)):
    return service(request).profile(current_user)


@router.get("/classroom", response_model=StudentClassroom)
def get_classroom(request: Request, current_user: UserPrincipal = Depends(student_user)):
    return service(request).classroom(current_user)


@router.get("/classmates", response_model=list[Classmate])
def get_classmates(request: Request, current_user: UserPrincipal = Depends(student_user)):
    return service(request).classmates(current_user)


@router.get("/teachers", response_model=list[TeacherSummary])
def get_teachers(request: Request, current_user: UserPrincipal = Depends(student_user)):
    return service(request).teachers(current_user)


@router.get("/activities", response_model=list[StudentActivity])
def get_activities(request: Request, current_user: UserPrincipal = Depends(student_user)):
    return service(request).activities(current_user)


@router.get("/activities/{activity_id}", response_model=StudentActivityDetail)
def get_activity_detail(activity_id: str, request: Request, current_user: UserPrincipal = Depends(student_user)):
    return service(request).activity_detail(activity_id, current_user)


@router.get("/activities/{activity_id}/export-pdf")
def export_activity_pdf(activity_id: str, request: Request, current_user: UserPrincipal = Depends(student_user)):
    pdf = service(request).export_pdf(activity_id, current_user)
    return Response(
        content=pdf,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="atividade-{activity_id}.pdf"'},
    )


@router.get("/grades", response_model=list[StudentPortalGrade])
def get_grades(request: Request, current_user: UserPrincipal = Depends(student_user)):
    return service(request).grades(current_user)


@router.get("/performance", response_model=StudentPortalPerformance)
def get_performance(request: Request, current_user: UserPrincipal = Depends(student_user)):
    return service(request).performance(current_user)
